### PR TITLE
es and es_en changes for unified models

### PR DESCRIPTION
**Pull request by mgrafu**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Have you signed your commits? Use ``git commit -s`` to sign.
- [ ] Do all unittests finish successfully before sending PR?
   1) ``pytest`` or (if your machine does not have GPU) ``pytest --cpu`` from the root folder (given you marked your test cases accordingly `@pytest.mark.run_only_on('CPU')`).
   2) Sparrowhawk tests ``bash tools/text_processing_deployment/export_grammars.sh --MODE=test ...``
- [ ] If you are adding a new feature: Have you added test cases for both `pytest` and Sparrowhawk [here](tests/nemo_text_processing).
- [ ] Have you added ``__init__.py`` for every folder and subfolder, including `data` folder which has .TSV files?
- [ ] Have you followed codeQL results and removed unused variables and imports (report is at the bottom of the PR in github review box) ?
- [ ] Have you added the correct license header `Copyright (c) 2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.` to all newly added Python files?
- [ ] If you copied [nemo_text_processing/text_normalization/en/graph_utils.py](nemo_text_processing/text_normalization/en/graph_utils.py) your header's second line should be `Copyright 2015 and onwards Google, Inc.`. See an example [here](https://github.com/NVIDIA/NeMo-text-processing/blob/main/nemo_text_processing/text_normalization/en/graph_utils.py#L2).
- [ ] Remove import guards (`try import: ... except: ...`) if not already done.
- [ ] If you added a new language or a new feature please update the [NeMo documentation](https://github.com/NVIDIA/NeMo/blob/main/docs/source/nlp/text_normalization/wfst/wfst_text_normalization.rst) (lives in different repo).
- [ ] Have you added your language support to [tools/text_processing_deployment/pynini_export.py](tools/text_processing_deployment/pynini_export.py).
  

**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation
- [ ] Test

If you haven't finished some of the above items you can still open "Draft" PR.

### CODE REVIEW

**Comment by ekmb** [comment]
@mgrafu is this PR WIP?

**Comment by mgrafu** [comment]
@ekmb yes! the majority of the changes are done, but I'm reviewing bugs before final review

File: nemo_text_processing/inverse_text_normalization/es/taggers/electronic.py [hunk 1]
```diff
@@ -26,22 +36,36 @@ class ElectronicFst(GraphFst):
...
+
+        alpha_num = (NEMO_ALPHA | num).optimize()
 
         symbols = pynini.string_file(get_abs_path("data/electronic/symbols.tsv")).invert()
+        if input_case == INPUT_CASED:
+            symbols = capitalized_input_graph(symbols)
 
         accepted_username = alpha_num | symbols
-        process_dot = pynini.cross("punto", ".")
+        dot = pynini.accep("punto")
```
Comment (tbartley94) [inline nemo_text_processing/inverse_text_normalization/es/taggers/electronic.py:62]: Could you explain this just so I'm sure I'm following?

File: nemo_text_processing/inverse_text_normalization/es/taggers/electronic.py [hunk 2]
```diff
@@ -51,25 +75,60 @@ def __init__(self):
...
+        if input_case == INPUT_CASED:
+            at |= pynini.accep("Arroba")
+
+        graph = username + delete_extra_space + pynutil.delete(at) + insert_space + delete_extra_space + domain_graph
 
         ############# url ###
-        protocol_end = pynini.cross(pynini.union("www", "w w w", "doble ve doble ve doble ve"), "www")
-        protocol_start = pynini.cross(pynini.union("http", "h t t p", "hache te te pe"), "http")
-        protocol_start |= pynini.cross(pynini.union("https", "h t t p s", "hache te te pe ese"), "https")
+        if input_case == INPUT_CASED:
```
Comment (tbartley94) [inline nemo_text_processing/inverse_text_normalization/es/taggers/electronic.py:110]: Question: do native speakers actually use Spanish letter names or is there a tendency to switch to English letters in this scenario?

File: nemo_text_processing/inverse_text_normalization/es/taggers/electronic.py [hunk 3]
```diff
@@ -15,8 +15,18 @@
 import pynini
```
Comment (tbartley94) [inline nemo_text_processing/inverse_text_normalization/es/taggers/electronic.py:15]: Does this graph work on Instagram/Twitter handles @your_example_here? If not, could you make the alteration so we have better compatibility with web scraping?

File: nemo_text_processing/inverse_text_normalization/es/taggers/money.py [hunk 4]
```diff
@@ -66,22 +86,31 @@ def __init__(self, cardinal: GraphFst, decimal: GraphFst):
 
         add_leading_zero_to_double_digit = (NEMO_DIGIT + NEMO_DIGIT) | (pynutil.insert("0") + NEMO_DIGIT)
 
+        one_graph = pynini.union("un", "una").optimize()
```
Comment (tbartley94) [inline nemo_text_processing/inverse_text_normalization/es/taggers/money.py:89]: Memory is blanking, but I recall that some currencies use `uno` in place of `un`. Could you confirm or deny?

File: nemo_text_processing/inverse_text_normalization/es/taggers/ordinal.py [hunk 5]
```diff
@@ -65,14 +79,20 @@ def __init__(self, cardinal: GraphFst):
...
+        self.graph_masc_num_no_exception = graph_o_suffix.optimize()
 
         # don't convert ordinals from one to nine inclusive
         graph_exception = pynini.project(pynini.union(graph_digit), 'input')
         graph_o_suffix = (pynini.project(graph_o_suffix, "input") - graph_exception.arcsort()) @ graph_o_suffix
         graph_a_suffix = (pynini.project(graph_a_suffix, "input") - graph_exception.arcsort()) @ graph_a_suffix
         graph_er_suffix = (pynini.project(graph_er_suffix, "input") - graph_exception.arcsort()) @ graph_er_suffix
 
+        if input_case == INPUT_CASED:
+            graph_exception = capitalized_input_graph(graph_exception)
```
Comment (tbartley94) [inline nemo_text_processing/inverse_text_normalization/es/taggers/ordinal.py:91]: Where is this graph_exception used again?

File: nemo_text_processing/inverse_text_normalization/es/taggers/time.py [hunk 6]
```diff
@@ -82,32 +88,59 @@ def __init__(self):
...
 
         # note that graph_hour will start from 2 hours
         # "1 o'clock" will be treated differently because it
         # is singular
         digits_2_to_23 = [str(digits) for digits in range(2, 24)]
         digits_1_to_59 = [str(digits) for digits in range(1, 60)]
 
-        graph_1oclock = pynini.cross("la una", "la 1")
-        graph_hour = pynini.cross("las ", "las ") + graph_1_to_100 @ pynini.union(*digits_2_to_23)
+        oneoclock = pynini.accep("la una")
```
Comment (tbartley94) [inline nemo_text_processing/inverse_text_normalization/es/taggers/time.py:99]: Use underscore `one_o_clock` This is going to cause variable errors if you spell this way.

File: nemo_text_processing/inverse_text_normalization/es/taggers/time.py [hunk 7]
```diff
@@ -82,32 +88,59 @@ def __init__(self):
...
+            graph_digit |= capitalized_input_graph(graph_digit).optimize()
+            graph_1_to_100 |= capitalized_input_graph(graph_1_to_100).optimize()
+            article |= pynini.accep("Las ").optimize()
+            half |= pynini.accep("Media").optimize()
+            quarter |= pynini.accep("Cuarto").optimize()
+            and_graph |= pynini.union("Y", "Con").optimize()
+            hours_word_graph |= pynini.accep(" Horas").optimize()
+            minutes_word_graph |= pynini.union(" Minuto", " Minutos").optimize()
+
+        graph_1oclock = pynini.cross(oneoclock, "la 1")
```
Comment (tbartley94) [inline nemo_text_processing/inverse_text_normalization/es/taggers/time.py:119]: `graph_1_o_clock`

File: nemo_text_processing/inverse_text_normalization/es/taggers/time.py [hunk 8]
```diff
@@ -127,10 +160,10 @@ def __init__(self):
         )
 
         # las nueve a eme (only convert on-the-hour times if they are followed by a suffix)
-        graph_1oclock_with_suffix = pynini.closure(pynini.accep("la "), 0, 1) + pynini.cross("una", "1")
-        graph_hour_with_suffix = pynini.closure(pynini.accep("las "), 0, 1) + graph_1_to_100 @ pynini.union(
-            *digits_2_to_23
+        graph_1oclock_with_suffix = pynini.closure(pynini.union("la ", "La "), 0, 1) + pynini.cross(
+            pynini.union("una", "Unia"), "1"
```
Comment (tbartley94) [inline nemo_text_processing/inverse_text_normalization/es/taggers/time.py:164]: `Unia`?

File: nemo_text_processing/inverse_text_normalization/es/taggers/time.py [hunk 9]
```diff
@@ -171,14 +204,14 @@ def __init__(self):
...
-                pynini.cross("cinco", "55"),
-                pynini.cross("diez", "50"),
-                pynini.cross("cuarto", "45"),
-                pynini.cross("veinte", "40"),
-                pynini.cross("veinticinco", "30"),
+                pynini.cross(pynini.union("cinco", "Cinco"), "55"),
+                pynini.cross(pynini.union("diez", "Diez"), "50"),
+                pynini.cross(pynini.union("cuarto", "Cuarto"), "45"),
+                pynini.cross(pynini.union("veinte", "Veinte"), "40"),
+                pynini.cross(pynini.union("veinticinco", "Veinticinco"), "35"),
```
Comment (tbartley94) [inline nemo_text_processing/inverse_text_normalization/es/taggers/time.py:214]: Also add the informal version `veinte y cinco`. It's probably unnecessary but no harm for ITN.

File: nemo_text_processing/inverse_text_normalization/es/verbalizers/time.py [hunk 10]
```diff
@@ -36,11 +36,13 @@ def __init__(self):
         add_leading_zero_to_double_digit = (NEMO_DIGIT + NEMO_DIGIT) | (pynutil.insert("0") + NEMO_DIGIT)
 
         # hour may or may not include preposition ("la" or "las")
+        preposition = pynini.union("la ", "las ", "La ", "Las ")
```
Comment (tbartley94) [inline nemo_text_processing/inverse_text_normalization/es/verbalizers/time.py:39]: `article` not `preposition` (latter more confusing)

File: nemo_text_processing/inverse_text_normalization/es_en/data/es_whitelist.tsv [hunk 11]
```diff
@@ -1,16 +1,16 @@
-ud.	usted
```
Comment (tbartley94) [inline nemo_text_processing/inverse_text_normalization/es_en/data/es_whitelist.tsv:1]: Same as previous comment for casing

File: nemo_text_processing/inverse_text_normalization/es_en/taggers/tokenize_and_classify.py [hunk 12]
```diff
@@ -151,7 +153,7 @@ def __init__(
                 | pynutil.add_weight(en_money_graph, 1.1)
                 | pynutil.add_weight(telephone_graph, 1.6)
                 | pynutil.add_weight(en_telephone_graph, 1.1)
-                | pynutil.add_weight(electronic_graph, 1.6)
+                | pynutil.add_weight(electronic_graph, 2.3)
```
Comment (tbartley94) [inline nemo_text_processing/inverse_text_normalization/es_en/taggers/tokenize_and_classify.py:156]: Pretty high weight, no?

File: nemo_text_processing/text_normalization/es/graph_utils.py [hunk 13]
```diff
@@ -40,6 +40,10 @@
 fem_hundreds = hundreds @ pynini.cdrewrite(pynini.cross("ientos", "ientas"), "", "", NEMO_SIGMA)
 
 
+ES_MINUS = pynini.union("menos", "Menos", "MENOS").optimize()
```
Comment (tbartley94) [inline nemo_text_processing/text_normalization/es/graph_utils.py:43]: Any informal form of these that we should consider?

**Review by tbartley94 — CHANGES_REQUESTED** [review]
Left comments and requested changes. There's a few mispellings, questions about edge cases, and request for variable renaming.

I trust the test cases are boilerplate, didn't go into much detail there.

File: nemo_text_processing/inverse_text_normalization/es/data/dates/year_suffix_cased.tsv [hunk 14]
```diff
@@ -0,0 +1,11 @@
+A. N. E.	antes de nuestra era
```
[the remark this answers is not in the document]
Comment (mgrafu) [reply nemo_text_processing/inverse_text_normalization/es/data/dates/year_suffix_cased.tsv:1]: this is Spanish, not Latin :)

[the remark this answers is not in the document]
Comment (tbartley94) [reply nemo_text_processing/inverse_text_normalization/es/data/dates/year_suffix_cased.tsv:1]: Yes, but the Latin will show up in English speech, so I would still expect normalization of `anno domini` if I'm running the normalizer over English text. I'm curious if the Latin similarly shows up in Spanish speech.

File: nemo_text_processing/text_normalization/es/graph_utils.py (hunk 13 above)
Comment (mgrafu) [reply nemo_text_processing/text_normalization/es/graph_utils.py:43]: should be fine as is -- especially since this PR is focused more on adding coverage of capitalization specifically and not necessarily general improvements :)

File: nemo_text_processing/inverse_text_normalization/es_en/taggers/tokenize_and_classify.py (hunk 12 above)
Comment (mgrafu) [reply nemo_text_processing/inverse_text_normalization/es_en/taggers/tokenize_and_classify.py:156]: yep -- language mix requires this, otherwise we get chaos

File: nemo_text_processing/inverse_text_normalization/es_en/data/es_whitelist.tsv (hunk 11 above)
Comment (mgrafu) [reply nemo_text_processing/inverse_text_normalization/es_en/data/es_whitelist.tsv:1]: preferred output would be capitalized always

File: nemo_text_processing/inverse_text_normalization/es/taggers/time.py (hunk 9 above)
Comment (mgrafu) [reply nemo_text_processing/inverse_text_normalization/es/taggers/time.py:214]: probably better not to, considering that we don't want erroneous ITN for an input like "Of these apples I got twenty and five of the other ones" -- sounds odd in English, but likely to exist in Spanish

File: nemo_text_processing/inverse_text_normalization/es/taggers/time.py (hunk 6 above)
Comment (mgrafu) [reply nemo_text_processing/inverse_text_normalization/es/taggers/time.py:99]: why would it cause variable errors?

File: nemo_text_processing/inverse_text_normalization/es/taggers/money.py (hunk 4 above)
Comment (mgrafu) [reply nemo_text_processing/inverse_text_normalization/es/taggers/money.py:89]: can confirm this case should not be seen by the grammar (unless ASR is having significant issues)

File: nemo_text_processing/inverse_text_normalization/es/taggers/electronic.py (hunk 1 above)
Comment (mgrafu) [reply nemo_text_processing/inverse_text_normalization/es/taggers/electronic.py:62]: graph can convert either 'punto' or (if capitalized) 'Punto' to '.' in these contexts

File: nemo_text_processing/inverse_text_normalization/es/taggers/electronic.py (hunk 3 above)
Comment (mgrafu) [reply nemo_text_processing/inverse_text_normalization/es/taggers/electronic.py:15]: very valid use case! but this PR is specifically to enable capitalized inputs -- we can add that as a task in the backlog for when we can schedule general improvements

File: nemo_text_processing/inverse_text_normalization/es/taggers/cardinal.py [hunk 15]
```diff
@@ -161,18 +174,35 @@ def __init__(self):
...
-            pynutil.insert("negative: ") + pynini.cross("menos", "\"-\"") + NEMO_SPACE, 0, 1
+            pynutil.insert("negative: ") + pynini.cross(ES_MINUS, "\"-\"") + NEMO_SPACE, 0, 1
         )
 
         final_graph = optional_minus_graph + pynutil.insert("integer: \"") + self.graph + pynutil.insert("\"")
 
         final_graph = self.add_tokens(final_graph)
         self.fst = final_graph.optimize()
+
+    def delete_word(self, word: str):
```
[the remark this answers is not in the document]
Comment (mgrafu) [reply nemo_text_processing/inverse_text_normalization/es/taggers/cardinal.py:201]: i agree, but it exists hear for consistency with the work done in EN since both graphs will work together -- would it make more sense to directly import from the english graph instead, or still in utils?

File: nemo_text_processing/inverse_text_normalization/es/taggers/cardinal.py [hunk 16]
```diff
@@ -161,18 +174,35 @@ def __init__(self):
...
 
         # don't convert cardinals from zero to nine inclusive
         graph_exception = pynini.project(pynini.closure(NEMO_SPACE, 0, 1) + (graph_digit | graph_zero), 'input')
 
-        self.graph = (pynini.project(graph, "input") - graph_exception.arcsort()) @ graph
+        self.graph = ((pynini.project(graph, "input") - graph_exception.arcsort()) @ graph).optimize()
 
         optional_minus_graph = pynini.closure(
-            pynutil.insert("negative: ") + pynini.cross("menos", "\"-\"") + NEMO_SPACE, 0, 1
+            pynutil.insert("negative: ") + pynini.cross(ES_MINUS, "\"-\"") + NEMO_SPACE, 0, 1
```
[the remark this answers is not in the document]
Comment (mgrafu) [reply nemo_text_processing/inverse_text_normalization/es/taggers/cardinal.py:193]: yes -- in different capitalizations (or forms if we find that necessary in the future)

File: nemo_text_processing/inverse_text_normalization/es/data/whitelist.tsv [hunk 17]
```diff
@@ -1,16 +1,47 @@
-ud.	usted
-uds.	ustedes
-vd.	vosotros
-vds.	vosotros
-dr.	doctor
-dra.	doctora
-d.	don
-da.	doña
-ee. uu.	estados unidos
+Ud.	usted
```
[the remark this answers is not in the document]
Comment (mgrafu) [reply nemo_text_processing/inverse_text_normalization/es/data/whitelist.tsv:1]: answered below instead

File: nemo_text_processing/inverse_text_normalization/es/data/roman/hundreds.tsv [hunk 18]
```diff
@@ -1,9 +1,9 @@
-c	1
```
[the remark this answers is not in the document]
Comment (mgrafu) [reply nemo_text_processing/inverse_text_normalization/es/data/roman/hundreds.tsv:1]: ITN, so unless the input can be distinguished, we can have just one output -- all caps would be preferred in that case

File: nemo_text_processing/inverse_text_normalization/es/data/dates/year_suffix_cased.tsv (hunk 14 above)
[the remark this answers is not in the document]
Comment (mgrafu) [reply nemo_text_processing/inverse_text_normalization/es/data/dates/year_suffix_cased.tsv:1]: unlikely to show up, tbh -- AC/DC is vastly more popular, so I would sat _if it did_ , better not to denorm

[the remark this answers is not in the document]
Comment (tbartley94) [reply nemo_text_processing/inverse_text_normalization/es/data/dates/year_suffix_cased.tsv:1]: Gotcha!

File: nemo_text_processing/inverse_text_normalization/es/data/roman/digit.tsv [hunk 19]
```diff
@@ -1,9 +1,9 @@
...
-ii	2
-iii	3
-iv	4
-v	5
-vi	6
-vii	7
-viii	8
-ix	9
\ No newline at end of file
+I	1
```
[the remark this answers is not in the document]
Comment (mgrafu) [reply nemo_text_processing/inverse_text_normalization/es/data/roman/digit.tsv:1]: you can test it out :) we did have to add some exceptions for words like 'mi', 'vi', and 'di'

File: nemo_text_processing/inverse_text_normalization/es/taggers/cardinal.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (mgrafu) [reply nemo_text_processing/inverse_text_normalization/es/taggers/cardinal.py:201]: @tbartley94 checking in on this since merge should happen soon

File: nemo_text_processing/inverse_text_normalization/es/taggers/ordinal.py (hunk 5 above)
Comment (mgrafu) [reply nemo_text_processing/inverse_text_normalization/es/taggers/ordinal.py:91]: removing

File: nemo_text_processing/inverse_text_normalization/es/taggers/time.py (hunk 6 above)
Comment (mgrafu) [reply nemo_text_processing/inverse_text_normalization/es/taggers/time.py:99]: changed

File: nemo_text_processing/inverse_text_normalization/es/taggers/time.py (hunk 7 above)
Comment (mgrafu) [reply nemo_text_processing/inverse_text_normalization/es/taggers/time.py:119]: changed as above instead

File: nemo_text_processing/inverse_text_normalization/es/taggers/time.py (hunk 8 above)
Comment (mgrafu) [reply nemo_text_processing/inverse_text_normalization/es/taggers/time.py:164]: typo fixed! good eye

File: nemo_text_processing/inverse_text_normalization/es/verbalizers/time.py (hunk 10 above)
Comment (mgrafu) [reply nemo_text_processing/inverse_text_normalization/es/verbalizers/time.py:39]: where was my brain with this one -- changed

File: nemo_text_processing/inverse_text_normalization/es/taggers/cardinal.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (tbartley94) [reply nemo_text_processing/inverse_text_normalization/es/taggers/cardinal.py:201]: Directly import from english graph and leave an issue for it to be moved in general. We want to minimzie duplicate code.

File: nemo_text_processing/inverse_text_normalization/es/taggers/electronic.py (hunk 3 above)
Comment (tbartley94) [reply nemo_text_processing/inverse_text_normalization/es/taggers/electronic.py:15]: Could you please add it as a future issue to keep consistency with other graphs being worked on?

File: nemo_text_processing/inverse_text_normalization/es/taggers/ordinal.py (hunk 5 above)
Comment (tbartley94) [reply nemo_text_processing/inverse_text_normalization/es/taggers/ordinal.py:91]: Not seeing the commit

File: nemo_text_processing/inverse_text_normalization/es/taggers/time.py (hunk 6 above)
Comment (tbartley94) [reply nemo_text_processing/inverse_text_normalization/es/taggers/time.py:99]: Murphy's law, someone can speed read and see `onoclock` `oneclock`. Readibility save lines.

File: nemo_text_processing/inverse_text_normalization/es/taggers/cardinal.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (mgrafu) [reply nemo_text_processing/inverse_text_normalization/es/taggers/cardinal.py:201]: nm, functionality is slightly different -- we can open an issue to track adjustment but may not make sense to change EN graph now because it could delay the availability of this one

File: nemo_text_processing/inverse_text_normalization/es/taggers/time.py (hunk 6 above)
Comment (mgrafu) [reply nemo_text_processing/inverse_text_normalization/es/taggers/time.py:99]: true words are true

File: nemo_text_processing/inverse_text_normalization/es/taggers/ordinal.py [hunk 20]
```diff
@@ -88,7 +88,6 @@ def __init__(self, cardinal: GraphFst, input_case: str = INPUT_LOWER_CASED):
         graph_er_suffix = (pynini.project(graph_er_suffix, "input") - graph_exception.arcsort()) @ graph_er_suffix
 
         if input_case == INPUT_CASED:
-            graph_exception = capitalized_input_graph(graph_exception)
```
Comment (tbartley94) [inline nemo_text_processing/inverse_text_normalization/es/taggers/ordinal.py]: Was this moved or just improved naming?

File: nemo_text_processing/inverse_text_normalization/es/taggers/telephone.py [hunk 21]
```diff
@@ -18,7 +18,6 @@
 from nemo_text_processing.text_normalization.en.graph_utils import (
     INPUT_CASED,
     INPUT_LOWER_CASED,
-    MIN_NEG_WEIGHT,
```
Comment (tbartley94) [inline nemo_text_processing/inverse_text_normalization/es/taggers/telephone.py]: Where was this moved?

**Review by tbartley94 — COMMENTED** [review]
Waiting on merge conflicts to be resolved then will approve

File: nemo_text_processing/inverse_text_normalization/es/taggers/telephone.py (hunk 21 above)
Comment (mgrafu) [reply nemo_text_processing/inverse_text_normalization/es/taggers/telephone.py]: this was flagged as an unused import, so removed

File: nemo_text_processing/inverse_text_normalization/es/taggers/ordinal.py (hunk 20 above)
Comment (mgrafu) [reply nemo_text_processing/inverse_text_normalization/es/taggers/ordinal.py]: unused

**Review by tbartley94 — APPROVED** [review]
LGTM
